java-checks/src/main/java/org/sonar/java/checks/PrivateFieldUsedLocallyCheck.java

```diff
@@ -0,0 +1,221 @@
+      if (fieldsLiveInMethodEntry.contains(privateFieldSymbol)) {
+        return;
+      }
+    }
+
+    reportIssue(((VariableTree) privateFieldSymbol.declaration()).simpleName(), String.format(MESSAGE, privateFieldSymbol.name()));
+  }
+
+  private void initMethodsLiveFields(TypeSymbol classSymbol) {
+    fieldsLiveInMethodEntryPerMethod = new HashMap<>();
+    fieldsReadOnAnotherInstance = new HashSet<>();
+
+    for (Symbol memberSymbol : classSymbol.memberSymbols()) {
+
+      if (memberSymbol.isMethodSymbol() && memberSymbol.declaration() != null) {
+        MethodTree methodTree = (MethodTree) memberSymbol.declaration();
+
+        if (methodTree.block() != null) {
+          CFG cfg = CFG.build(methodTree);
+          LiveFields liveFields = LiveFields.analyze(cfg);
+          fieldsLiveInMethodEntryPerMethod.put(methodTree, liveFields.getIn(cfg.entry()));
+          fieldsReadOnAnotherInstance.addAll(liveFields.fieldsReadOnAnotherInstance);
+        }
+      }
+    }
+  }
+
+  private static boolean isPrivateField(Symbol memberSymbol) {
+    return memberSymbol.isPrivate() && memberSymbol.isVariableSymbol();
+  }
+
+  private static class LiveFields extends LiveVariables {
+
+    private Set<Symbol> fieldsReadOnAnotherInstance = new HashSet<>();
+
+    private LiveFields(CFG cfg) {
+      super(cfg);
+    }
+
+    public static LiveFields analyze(CFG cfg) {
+      return getLiveVariables(new LiveFields(cfg));
+    }
+
+    private void checkFieldReadOnAnotherInstance(Tree element) {
+      if (element.is(Kind.MEMBER_SELECT)) {
+        MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) element;
+
+        if (memberSelect.expression().is(Kind.IDENTIFIER)) {
+          String objectName = ((IdentifierTree) memberSelect.expression()).name();
+          Symbol symbol = memberSelect.identifier().symbol();
+
+          if (!symbol.isStatic() && !"this".equals(objectName) && isField(symbol)) {
+            fieldsReadOnAnotherInstance.add(symbol);
+          }
+        }
+      }
+    }
+
+    @Override
+    protected void processBlockElement(Tree element, Set<Symbol> blockKill, Set<Symbol> blockGen, Set<Tree> assignmentLHS) {
+      switch (element.kind()) {
+        case ASSIGNMENT:
+          processAssignment((AssignmentExpressionTree) element, blockKill, blockGen, assignmentLHS);
+          break;
+        case IDENTIFIER:
+        case MEMBER_SELECT:
+          Symbol symbol = getFieldSymbol((ExpressionTree) element);
+          if (symbol != null && !assignmentLHS.contains(element)) {
+            blockGen.add(symbol);
+          }
+          break;
+        default:
+          // Ignore other kind of elements, no change of gen/kill
+      }
+
+      checkFieldReadOnAnotherInstance(element);
+    }
+
+    private void processAssignment(AssignmentExpressionTree assignmentExpression, Set<Symbol> blockKill, Set<Symbol> blockGen, Set<Tree> assignmentLHS) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Rename this method; there is a "private" method in the parent class with the same name. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2177)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make "processAssignment" a "static" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)